predicators/approaches/grammar_search_invention_approach.py

```diff
@@ -513,56 +572,135 @@ class _FeatureDiffInequalitiesPredicateGrammar(
+            for ret_val in self._yield_pred_given_const(
+                    feature_ranges, constant_idx, constant, cost):
+                yield ret_val
+
+
+@dataclass(frozen=True, eq=False, repr=False)
+class _EuclideanDistancePredicateGrammar(
+        _SingleFeatureInequalitiesPredicateGrammar):
+    """Generates predicates of the form "|0.x - 1.x|^2 + |0.y - 1.y|^2 <= c^2".
+    Importantly, this only operates over types that have features
+    named "x" and "y".
+    """
+
+    def _compute_xy_bounds(self, feature_ranges: Dict[Type,
+                                                      Dict[str, Tuple[float,
+                                                                      float]]],
+                           t1: Type, t2: Type) -> Tuple[float, float]:
+        # To create our classifier, we need to leverage the
+        # upper and lower bounds of its x, y features.
+        lbx1, ubx1 = feature_ranges[t1]["x"]
+        lbx2, ubx2 = feature_ranges[t2]["x"]
+        lby1, uby1 = feature_ranges[t1]["y"]
+        lby2, uby2 = feature_ranges[t2]["y"]
+        # Compute the upper and lower bounds of each feature range.
+        lbx, ubx = utils.compute_abs_bounds_given_frange(
+            lbx1, ubx1, lbx2, ubx2)
+        lby, uby = utils.compute_abs_bounds_given_frange(
+            lby1, uby1, lby2, uby2)
+        # Now, use these to compute the upper and lower bounds of
+        # the squared expression of interest.
+        lb = lbx**2 + lby**2
+        ub = ubx**2 + uby**2
+        return (lb, ub)
+
+    def _generate_pred_given_constant(self, constant_idx: int, constant: float,
+                                      t1: Type, t2: Type) -> Predicate:
+        # Create classifier.
+        comp, comp_str = le, "<="
+        diff_classifier = _EuclideanAttributeDiffCompareClassifier(
+            0, t1, "x", 1, t2, "y", constant, constant_idx, comp, comp_str)
+        name = str(diff_classifier)
+        types = [t1, t2]
+        pred = Predicate(name, types, diff_classifier)
+        return pred
+
+    def enumerate(self) -> Iterator[Tuple[Predicate, float]]:
+        # Get ranges of feature values from data.
+        feature_ranges = self._get_feature_ranges()
+        # Edge case: if there are no features at all, return immediately.
+        if not any(r for r in feature_ranges.values()):
+            return
+
+        # Start by generating predicates with a very small constant,
+        # to indicate that the objects are touching/overlapped.
+        for (t1, t2) in itertools.combinations_with_replacement(
+                sorted(self.types), 2):
+            if t1 == t2:
+                continue
+            if ("x" in t1.feature_names and "x" in t2.feature_names
+                    and "y" in t1.feature_names and "y" in t2.feature_names):
+                lb, ub = self._compute_xy_bounds(feature_ranges, t1, t2)
+                constant = ((ub - lb) / 500) + lb
```

Comment:
500 also looks magic to me

Reply:
it is! will pull these out and make any magic numbers global constants